File: algorithms/[692]前K个高频单词/solution.ts
// topics = ["数组"]

/**
 * 哈希表 & 排序
 * time O(nlogn), space O(n), n 为数组的长度
 */
function topKFrequent(words: string[], k: number): string[] {
  const cnt = new Map<string, number>();
  for (const word of words) {
    cnt.set(word, (cnt.get(word) ?? 0) + 1);
  }

  const rec: string[] = [];
  for (const entry of cnt.keys()) {
    rec.push(entry);
  }

  rec.sort((word1, word2) => {
    return cnt.get(word1) === cnt.get(word2) ? word1.localeCompare(word2) : cnt.get(word2)! - cnt.get(word1)!;
  });

  return rec.slice(0, k);
}

File: algorithms/[692]前K个高频单词/solution.py
# topics = ["数组", "堆"]

from __future__ import annotations
from collections import Counter
from typing import Dict, List
import heapq


class Word:
    def __init__(self, word: str, fre: int):
        self.word = word
        self.fre = fre

    def __lt__(self, other: Word):
        if self.fre != other.fre:
            return self.fre < other.fre
        return self.word > other.word


class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        """
        Hash Table & Heap
        time O(nlogk), space O(n), n 为数组的长度
        近似题 1738.找出第 K 大的异或坐标值
        """
        # 计数器
        c: Dict[str, int] = Counter(words)
        # 最小堆
        h: List[Word] = []

        for word, cnt in c.items():
            if len(h) < k:
                heapq.heappush(h, Word(word, cnt))
            else:
                heapq.heappushpop(h, Word(word, cnt))

        h.sort(reverse=True)
        return [w.word for w in h]
